core/src/main/java/com/linecorp/armeria/common/stream/InputStreamStreamMessage.java

```diff
@@ -0,0 +1,370 @@
+
+    InputStreamStreamMessage(InputStream inputStream, @Nullable ExecutorService blockingTaskExecutor,
+                             int bufferSize) {
+        requireNonNull(inputStream, "inputStream");
+        this.inputStream = inputStream;
+        this.blockingTaskExecutor = blockingTaskExecutor;
+        this.bufferSize = bufferSize;
+    }
+
+    @Override
+    public ByteStreamMessage range(long offset, long length) {
+        checkArgument(offset >= 0, "offset: %s (expected: >= 0)", offset);
+        checkArgument(length > 0, "length: %s (expected: > 0)", length);
+        checkState(subscribed == 0, "cannot specify range(%s, %s) after this %s is subscribed",
+                   offset, length, InputStreamStreamMessage.class);
+        this.offset = offset;
+        this.length = length;
+        return this;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return !completionFuture.isDone();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (isOpen()) {
+            return false;
+        }
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        return inputStreamSubscription == null || inputStreamSubscription.position == 0;
+    }
+
+    @Override
+    public long demand() {
+        final InputStreamSubscription inputStreamSubscription = this.inputStreamSubscription;
+        if (inputStreamSubscription != null) {
+            return inputStreamSubscription.requested;
+        } else {
+            return 0;
+        }
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return completionFuture;
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        if (!subscribedUpdater.compareAndSet(this, 0, 1)) {
+            subscriber.onSubscribe(NoopSubscription.get());
+            subscriber.onError(new IllegalStateException("Only single subscriber is allowed!"));
+            return;
+        }
+
+        if (executor.inEventLoop()) {
+            subscribe0(subscriber, executor, options);
+        } else {
+            executor.execute(() -> subscribe0(subscriber, executor, options));
+        }
+    }
+
+    private void subscribe0(Subscriber<? super HttpData> subscriber, EventExecutor executor,
+                            SubscriptionOption... options) {
+        final ExecutorService blockingTaskExecutor;
+        if (this.blockingTaskExecutor != null) {
+            blockingTaskExecutor = this.blockingTaskExecutor;
+        } else {
+            final ServiceRequestContext serviceRequestContext = ServiceRequestContext.currentOrNull();
+            if (serviceRequestContext != null) {
+                blockingTaskExecutor = serviceRequestContext.blockingTaskExecutor();
+            } else {
+                blockingTaskExecutor = CommonPools.blockingTaskExecutor();
+            }
+        }
```

Comment:
I see some duplications between this class and `PathStreamMessage`. Can we deduplicate as much as possible, potentially introducing a common supertype?

Reply:
> Can we deduplicate as much as possible, potentially introducing a common `supertype`?

SGTM! but may I handle it with another issue & PR?
Cause I think we need some discussion about this issue 🙇

Reply:
If you want to handle it in a separate issue, it is okay to send a clean-up PR.

Reply:
I'll send clean-up PR after this PR is merged~!